src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspacePullDiagnosticsHandler.cs

```diff
@@ -22,12 +24,15 @@ internal abstract class AbstractWorkspacePullDiagnosticsHandler<TDiagnosticsPara
     protected readonly IDiagnosticSourceManager DiagnosticSourceManager;
 
     /// <summary>
-    /// Flag that represents whether the LSP view of the world has changed.
-    /// It is totally fine for this to somewhat over-report changes
-    /// as it is an optimization used to delay closing workspace diagnostic requests
-    /// until something has changed.
+    /// Gate to guard access to <see cref="_categoryToLspChanged"/>
     /// </summary>
-    private int _lspChanged = 0;
+    private readonly object _gate = new();
+
+    /// <summary>
+    /// Stores the LSP changed state on a per category basis.  This ensures that requests for different categories
+    /// are 'walled off' from each other and only reset state for their own category.
+    /// </summary>
+    private readonly Dictionary<string, bool> _categoryToLspChanged = new();
```

Comment:
❗ If we make this a `Dictionary<string, CancellationTokenSeries>`, we can get rid of polling for changes altogether.

Reply:
Actually, with `CancellationTokenSeries` as the value, I believe we can move to `ImmutableDictionary<string, CancellationTokenSeries>` and also get rid of the gate.

Reply:
@sharwell I'm not seeing an easy way to use `CancellationSeries` here - I was initially thinking something like

```
private void UpdateLspChanged()
{
      // Loop through our map of source -> has changed and mark them as all having changed.
      foreach (var category in _categoryToLspChanged.Keys.ToImmutableArray())
      {
          _categoryToLspChanged[category].Token.Cancel()?
      }
}

protected async Task WaitForChangesAsync(...)
{
    var token = _categoryToLspChanged[category].Token?;
  
    // Wait for token to be cancelled
    ...
  
    _categoryToLspChanged[category].CreateNext();
}
```

but cancellation series doesn't seem to provide access to the current cancellation token.  I can't have `UpdateLspChanged` create the next item in the cancellation series because the change event could happen before the `WaitForChangesAsync` method gets called, so it gets the new token that hasn't been cancelled.

Reply:
I implemented a version of this for review in 82dde3a5007e861e6a434f1c40c2d3ca31af62c1